File: src/opt.c
/* -*- tab-width : 2 -*- */
#include "opt.h"

void free_opts(struct opts* opt) {
  while(opt) {
    void* tmp=opt->next;
    if(opt->value) {
      dealloc((void*)opt->value);
      opt->value=NULL;
    }
    if(opt->name) {
      dealloc((void*)opt->name);
      opt->name=NULL;
    }

    opt->next=NULL;
    tmp=opt;
    opt=opt->next;
    dealloc(tmp);
  }
}

char* sexp_opts(struct opts* opt) {
  void* ret=q("");
  for(;opt;opt=opt->next)
    ret=s_cat(ret,q("(\""),q(opt->name),q("\"\""),escape_string((char*)opt->value),q("\")"),NULL);
  return ret;
}

struct opts* load_opts(const char* path) {
  FILE* fp;
  char buf[1024];
  struct opts opt;
  struct opts *cur=&opt;
  opt.next=NULL;

  if((fp=fopen(path,"r"))==NULL)
    return NULL;

  while(fgets(buf,1024,fp) !=NULL) {
    int i,mode,last;
    cur->next=(struct opts*)alloc(sizeof(struct opts));
    cur=cur->next;
    cur->type=OPT_VOID;
    cur->value=NULL;
    cur->name=NULL;
    cur->next=NULL;
    for(i=0,mode=0,last=0;i<1024&&buf[i]!='\0';++i) {
      if(buf[i]=='\t'||buf[i]=='\n') {
        switch (mode++) {
        case 0:
          cur->name=subseq(buf,last,i);
          break;
        case 1:
          cur->type=buf[i-1]-'0';
          break;
        case 2:
          cur->value=subseq(buf,last,i);
          break;
        }
        last=i+1;
      }
    }
  }
  fclose (fp);
  return opt.next;
}

int save_opts(const char* path,struct opts* opt) {
  FILE* fp;

  if((fp=fopen(path,"w"))==NULL)
    return 0;

  while(opt) {
    fprintf(fp,"%s\t%d\t%s\n",opt->name,opt->type,opt->value);
    opt=opt->next;
  }
  fclose(fp);
  return 1;
}

int set_opt(struct opts** opts,const char* name,char* value) {
  int found=0;
  struct opts* opt=*opts;

  while(opt) {
    if(strcmp(opt->name,name)==0) {
      found=1;
      opt->value=remove_char("\n\t",value);
    }
    opt=opt->next;
  }
  if(!found) {
    opt=(struct opts*)alloc(sizeof(struct opts));
    opt->next=*opts;
    opt->type=0;
    opt->name=(const char*)remove_char("\n\t",(char*)name);
    opt->value=remove_char("\n\t",value);
    *opts=opt;
  }
  return 1;
}

char* _get_opt(struct opts* opt,const char* name) {
  for(;opt;opt=opt->next)
    if(strcmp(opt->name,name)==0)
      return (char*)opt->value;
  return NULL;
}

char* _getenv(const char* name) {
  char* name_=substitute_char('_','.',q(name));
  char* ret=getenv(name_);
  s(name_);
  return ret;
}

char* get_opt(const char* name,int env) {
  char* ret=NULL;
  if(env)ret=_getenv(name);
  if(!ret) {
    ret=_get_opt(local_opt,name);
    if(!ret)
      ret=_get_opt(global_opt,name);
  }
  return ret;
}

int unset_opt(struct opts** opts,const char* name) {
  struct opts *opt=*opts;
  struct opts dummy;
  struct opts *before=&dummy;
  before->next=opt;
  while(opt) {
    if(strcmp(opt->name,name)==0) {
      before->next=opt->next;
      opt->next=NULL;
      free_opts(opt);
      opt=before;
    }
    before=opt;
    opt=opt->next;
  }
  *opts=dummy.next;
  return 1;
}

LVal add_command(LVal cmd,const char* name,const char* short_name,sub_command_fnc call,int show_opt,int terminating) {
  struct sub_command* p=alloc(sizeof(struct sub_command));
  p->name=name?q(name):NULL;
  p->short_name=short_name?q(short_name):NULL;
  p->call=call;
  p->show_opt=show_opt;
  p->terminating=terminating;
  cmd=cons(p,cmd);
  return cmd;
}

File: opt.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
)

var localOpt []opts
var globalOpt []opts

type subCommandFnc func(argv []string, cmd subCommand) int

type opts struct {
	name  string
	typ   int
	value string
}

type subCommand struct {
	name        string
	shortName   string
	call        subCommandFnc
	showOpt     int
	terminating bool
	description string
	argExample  string
}

type commandHelp struct {
	name     string
	usage    string
	commands []subCommand
	opts     []subCommand
	header   string
	footer   string
	call     subCommandFnc
}

func printOpts(opt []opts) {
	for _, op := range opt {
		fmt.Printf("%s=%s[]\n", op.name, op.value)
	}
}

func addHelp(base []commandHelp, name string, usage string, commands []subCommand,
	opts []subCommand, header string, footer string, call subCommandFnc) []commandHelp {
	return append(base, commandHelp{name: name, usage: usage,
		commands: commands, opts: opts, header: header, footer: footer, call: call})
}

func addCommand(base []subCommand, name string, shortName string, call subCommandFnc,
	showOpt int, terminating bool, description string, argExample string) []subCommand {

	return append(base, subCommand{name: name, shortName: shortName, call: call, showOpt: showOpt,
		terminating: terminating, description: description, argExample: argExample})
}

func setOpt(opt []opts, name string, value string, typ int) []opts {
	for _, v := range opt {
		if v.name == name {
			v.value = value
			return opt
		}
	}
	return append(opt, opts{name: name, value: value, typ: 0})
}

func _getOpt(opt []opts, name string) string {
	for _, v := range opt {
		if v.name == name {
			return v.value
		}
	}
	return ""
}

func unsetOpt(opt []opts, name string) []opts {
	for i, v := range opt {
		if v.name == name {
			return append(opt[0:i], opt[i+1:]...)
		}
	}
	return opt
}
func loadOpts(path string) (ret []opts) {
	f, err := os.Open(path)
	if err != nil {
		return []opts{}
	}
	defer f.Close()
	reader := bufio.NewReaderSize(f, 1024)
	for {
		cur := opts{}
		line, _, err := reader.ReadLine()
		buf := string(line)
		if buf != "" {
			for i, mode, last := 0, 0, 0; i < len(buf); i++ {
				if buf[i] == '\t' || buf[i] == '\n' || len(buf)-1 == i {
					mode++
					switch mode {
					case 1:
						cur.name = buf[last:i]
					case 2:
						//cur.typ = buf[i-1] - '0'
					case 3:
						cur.value = buf[last:]
					}
					last = i + 1
				}

			}
			ret = append(ret, cur)
		}
		if err == io.EOF {
			break
		} else if err != nil {
			panic(err)
		}
	}
	return
}

func saveOpts(path string, opt []opts) int {
	f, err := os.Create(path)
	if err != nil {
		return 0
	}
	defer f.Close()
	for _, op := range opt {
		f.WriteString(fmt.Sprintf("%s\t%d\t%s\n", op.name, 0, op.value))
	}
	return 1
}
